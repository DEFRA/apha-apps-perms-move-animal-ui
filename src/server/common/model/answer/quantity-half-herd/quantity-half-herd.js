import { RadioButtonAnswer } from '../radio-button/radio-button.js'

/** @import {RadioButtonConfig} from '../radio-button/radio-button.js' */

/**
 * export @typedef {'yes'|'no'|'unknown'} QuantityHalfHerdData
 * @typedef {{ quantityHalfHerd: QuantityHalfHerdData }} QuantityHalfHerdPayload
 */

/**
 * @augments {RadioButtonAnswer<QuantityHalfHerdPayload>}
 */
export class QuantityHalfHerdAnswer extends RadioButtonAnswer {
  /** @type {RadioButtonConfig} */
  static config = {
    payloadKey: 'quantityHalfHerd',
    options: {
      no: { label: 'No' },
      yes: { label: 'Yes' },
      unknown: { label: "I don't know" }
    },
    errors: {
      emptyOptionText:
        'Select if the number of cattle will be larger than half of the destination herd size'
    },
    layout: 'stacked'
  }
}
