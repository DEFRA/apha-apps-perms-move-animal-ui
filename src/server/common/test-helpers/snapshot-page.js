import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
