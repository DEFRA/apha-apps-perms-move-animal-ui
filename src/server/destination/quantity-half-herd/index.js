import { Page } from '../../common/model/page/page-model.js'

export class QuantityHalfHerdPage extends Page {
  urlPath = '/destination/quantity-half-herd'
}

export const quantityHalfHerdPage = new QuantityHalfHerdPage()
