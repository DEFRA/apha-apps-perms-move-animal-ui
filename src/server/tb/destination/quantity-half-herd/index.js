import { QuestionPageController } from '../../../common/controller/question-page-controller/question-page-controller.js'
import { QuantityHalfHerdAnswer } from '../../../common/model/answer/quantity-half-herd/quantity-half-herd.js'
import { QuestionPage } from '../../../common/model/page/question-page-model.js'
import { destinationSummaryPage } from '../summary/index.js'

export class QuantityHalfHerdPage extends QuestionPage {
  urlPath = '/destination/quantity-half-herd'
  sectionKey = 'destination'
  question =
    'Will the number of cattle be larger than half of the destination herd size?'

  questionKey = 'movingMoreThanHalfExistingHerd'

  Answer = QuantityHalfHerdAnswer

  nextPage() {
    return destinationSummaryPage
  }
}

export const quantityHalfHerdPage = new QuantityHalfHerdPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const quantityHalfHerd = new QuestionPageController(
  new QuantityHalfHerdPage()
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
