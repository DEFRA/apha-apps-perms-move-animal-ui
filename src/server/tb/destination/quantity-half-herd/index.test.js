import { quantityHalfHerdPage, QuantityHalfHerdPage } from './index.js'
import { describePageSnapshot } from '../../../common/test-helpers/snapshot-page.js'
import { destinationSummaryPage } from '../summary/index.js'
import { QuantityHalfHerdAnswer } from '../../../common/model/answer/quantity-half-herd/quantity-half-herd.js'

const sectionKey = 'destination'
const question =
  'Will the number of cattle be larger than half of the destination herd size?'
const questionKey = 'movingMoreThanHalfExistingHerd'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/destination/quantity-half-herd'

describe('QuantityHalfHerdPage', () => {
  let page

  beforeEach(() => {
    page = new QuantityHalfHerdPage()
  })

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(QuantityHalfHerdAnswer)
  })

  it('nextPage should return destination summary page', () => {
    const nextPage = page.nextPage()
    expect(nextPage).toBe(destinationSummaryPage)
  })

  it('should export page', () => {
    expect(quantityHalfHerdPage).toBeInstanceOf(QuantityHalfHerdPage)
  })

  describePageSnapshot({
    describes: 'QuantityHalfHerdPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
