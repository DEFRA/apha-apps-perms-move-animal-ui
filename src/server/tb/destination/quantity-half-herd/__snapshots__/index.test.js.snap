// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityHalfHerdPage QuantityHalfHerdPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
  <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
    <h1 class="govuk-fieldset__heading">
      Will the number of cattle be larger than half of the destination herd size?
    </h1>
  </legend>
  <div class="govuk-radios" data-module="govuk-radios">
    <div class="govuk-radios__item">
      <input class="govuk-radios__input" id="quantityHalfHerd" name="quantityHalfHerd" type="radio" value="no">
      <label class="govuk-label govuk-radios__label" for="quantityHalfHerd">
        No
      </label>
    </div>
    <div class="govuk-radios__item">
      <input class="govuk-radios__input" id="yes" name="quantityHalfHerd" type="radio" value="yes">
      <label class="govuk-label govuk-radios__label" for="yes">
        Yes
      </label>
    </div>
    <div class="govuk-radios__item">
      <input class="govuk-radios__input" id="unknown" name="quantityHalfHerd" type="radio" value="unknown">
      <label class="govuk-label govuk-radios__label" for="unknown">
        I don't know
      </label>
    </div>
  </div>
</fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
